### PR TITLE
feat: ifthenelse pypika [TCTC-3294]

### PR DESCRIPTION
**Pull request by Sanix-Darker**
## WHAT
- IfthenElse on Dates Step
- IfthenElse on Regex Step

### CODE REVIEW

File: server/src/weaverbird/backends/pypika_translator/operators.py [hunk 1]
```diff
@@ -5,6 +5,11 @@ class RegexOp(Enum):
     REGEXP = auto()
     SIMILAR_TO = auto()
     CONTAINS = auto()
+    REGEXP_LIKE = auto()
+    REGEXP_SUBSTR = auto()
+    REGEXP_EXTRACT = auto()
+    REGEXP_MATCH = auto()
+    REGEXP_CONTAINS = auto()
```
Comment (jGundermann) [inline server/src/weaverbird/backends/pypika_translator/operators.py:12]: These cases are not tested yet ?
Shouldn't we add the corresponding unit test ?

**Review by jGundermann — CHANGES_REQUESTED** [review]

File: server/src/weaverbird/backends/pypika_translator/operators.py (hunk 1 above)
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/operators.py:12]: Only `REGEXP_LIKE` is tested in this Pull Request, is it better for to remove the others ? 🤔 
Because i added them after checking some regex SQL syntax available !

Comment (jGundermann) [reply server/src/weaverbird/backends/pypika_translator/operators.py:12]: I mean making something similar to this :
https://github.com/ToucanToco/weaverbird/blob/master/server/tests/backends/sql_translator_unit_tests/test_filter_translators.py#L124

Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/operators.py]: Alright, i see, you mean something like this : https://github.com/ToucanToco/weaverbird/pull/1418/commits/e8c542b9f513e4f415570dff7b89ad129cbc78a5 ?

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 2]
```diff
@@ -1126,6 +1176,18 @@ def __init__(self, term: str | Field, date_format: str, alias: str | None = None
...
+class RegexpFunction(functions.Function):
+    REGEXP_LIKE = "REGEXP_LIKE"
+    REGEXP_CONTAINS = "REGEXP_CONTAINS"
+    NOT_REGEXP_LIKE = "NOT REGEXP_LIKE"
+    NOT_REGEXP_CONTAINS = "NOT REGEXP_CONTAINS"
+
+    def __init__(
+        self, keyword: str, term: str | Field, regexp_expression: str, alias: str | None = None
+    ) -> None:
+        super().__init__(keyword, term, regexp_expression, alias=alias)
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:1188]: Shouldn't `keyword` be a RegexOp ? Otherwise, I don't think this constructor is required, since it only calls `super().__init__`

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 3]
```diff
@@ -535,23 +533,52 @@ def _get_single_condition_criterion(
...
                     match self.REGEXP_OP:
                         case RegexOp.REGEXP:
                             return column_field.regexp(condition.value)
                         case RegexOp.SIMILAR_TO:
+
+                            compliant_regex = (
+                                compliant_regex.replace('%', '%%')
+                                if self.DIALECT in [SQLDialect.POSTGRES, SQLDialect.REDSHIFT]
+                                else compliant_regex
+                            )
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:548]: Why is this needed ?

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 4]
```diff
@@ -1126,6 +1176,18 @@ def __init__(self, term: str | Field, date_format: str, alias: str | None = None
         super().__init__("PARSE_DATE", term, date_format, alias=alias)
 
 
+class RegexpFunction(functions.Function):
+    REGEXP_LIKE = "REGEXP_LIKE"
+    REGEXP_CONTAINS = "REGEXP_CONTAINS"
+    NOT_REGEXP_LIKE = "NOT REGEXP_LIKE"
+    NOT_REGEXP_CONTAINS = "NOT REGEXP_CONTAINS"
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:1183]: I think this should live in the `RegexOp` enum, and it should be changed to `RegexOp(str, Enum)`

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 5]
```diff
@@ -562,16 +589,43 @@ def _get_single_condition_criterion(
                         case RegexOp.REGEXP:
                             return column_field.regexp(condition.value).negate()
                         case RegexOp.SIMILAR_TO:
+                            compliant_regex = (
+                                compliant_regex.replace('%', '%%')
+                                if self.DIALECT in [SQLDialect.POSTGRES, SQLDialect.REDSHIFT]
+                                else compliant_regex
+                            )
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:596]: If this is intended for escaping, it does not work, an `ESCAPE` statement must be used:

```
playground_db=# select "Transaction_date", "City" from sales WHERE "sales"."State" SIMILAR TO '%Aviv';
  Transaction_date   |   City
---------------------+----------
 2009-01-04 13:17:00 | Tel Aviv
(1 row)

playground_db=# select "Transaction_date", "City" from sales WHERE "sales"."State" SIMILAR TO '%%Aviv';
  Transaction_date   |   City
---------------------+----------
 2009-01-04 13:17:00 | Tel Aviv
(1 row)

playground_db=# select "Transaction_date", "City" from sales WHERE "sales"."State" SIMILAR TO '%Aviv' ESCAPE '%';
 Transaction_date | City
------------------+------
(0 rows)
```

**Review by jGundermann — APPROVED** [review]
Thx for the changes!

LGTM :cat:

**Review by lukapeschke — CHANGES_REQUESTED** [review]

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 4 above)
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py]: Make sense, updated here : https://github.com/ToucanToco/weaverbird/pull/1418/commits/748984ee43328c4f6d25ac0abaf82c1f2bf513a7

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 2 above)
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py]: True, the constructor is not needed !
For the name i rather prefer this class to stay as `RegexFunction` to not collide with the `RegexOp` class from `operators` what do you think ?
I updated that here : https://github.com/ToucanToco/weaverbird/pull/1418/commits/a3c4751870350c89f8da3617ccad6f4ea3dfc30d

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 6]
```diff
@@ -535,23 +533,52 @@ def _get_single_condition_criterion(
...
 
             case MatchCondition():
+                compliant_regex = _compliant_regex(condition.value)
 
                 if condition.operator == "matches":
                     match self.REGEXP_OP:
                         case RegexOp.REGEXP:
                             return column_field.regexp(condition.value)
                         case RegexOp.SIMILAR_TO:
+
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:543]: I feel like there should be an abstract method returning the match operator: Here, we're guessing what the child class needs and we have cases dependent on child classes (like on l.544, 561 or 572). I believe that it would lead to a better separation of concerns and easier to read code (and probably less LoC if we eliminate some cases).

We can keep this as is for this PR and do the refactoring later though

File: server/tests/backends/fixtures/filter/dates_from_until_pypika.json [hunk 7]
```diff
@@ -20,7 +20,7 @@
             {
               "column": "brewing_date",
               "operator": "until",
-              "value": "2022-01-07 00:00:00"
+              "value": "2022-01-07 01:00:00"
```
Comment (lukapeschke) [inline server/tests/backends/fixtures/filter/dates_from_until_pypika.json:23]: Why did the input have to be changed ? We should keep the same contract

File: server/tests/backends/fixtures/ifthenelse/dates_conditions_pypika.json [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            "column": "brewing_date",
+            "operator": "from",
+            "value": "2022-01-05 00:00:00"
+        },
+        "then": "'tintin'",
+        "else": {
+          "if": {
+              "column": "brewing_date",
+              "operator": "until",
+              "value": "2022-01-07 00:00:00"
```
Comment (lukapeschke) [inline server/tests/backends/fixtures/ifthenelse/dates_conditions_pypika.json:24]:
```suggestion
              "value": "2022-01-03 00:00:00"
```

We'll never get `tata` with a date later than the date in the first predicate

File: server/tests/backends/fixtures/filter/dates_from_until_pypika.json (hunk 7 above)
Comment (Sanix-Darker) [reply server/tests/backends/fixtures/filter/dates_from_until_pypika.json]: https://github.com/ToucanToco/weaverbird/pull/1418/commits/8d6d27c067adf502a2ea3f6bd619470e99c8f58f

File: server/src/weaverbird/backends/pypika_translator/operators.py [hunk 9]
```diff
@@ -1,10 +1,14 @@
 from enum import Enum, auto
 
 
-class RegexOp(Enum):
+class RegexOp(str, Enum):
     REGEXP = auto()
     SIMILAR_TO = auto()
     CONTAINS = auto()
```
[the remark this answers is not in the document]
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/operators.py]: Cleaned here : https://github.com/ToucanToco/weaverbird/pull/1418/commits/6ecbbd6737cfa6fbe9131eeb6b2a4853fb76b6a1

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 2 above)
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py]: removed here : https://github.com/ToucanToco/weaverbird/pull/1418/commits/d28d51bfb69073501f9bb829c725240ed122c65d

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 5 above)
Comment (lukapeschke) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py]: Thanks! I still don't understand why double `%` are required for postgres and redshift for `SIMILAR TO` though :thinking:

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 10]
```diff
@@ -683,10 +718,33 @@ def _get_single_condition_criterion(
...
-                if condition.operator == "until":
-                    return column_field <= condition.value
-                elif condition.operator == "from":
-                    return column_field >= condition.value
+
+                if isinstance(condition.value, (RelativeDate, datetime, str)):
+                    if isinstance(condition.value, RelativeDate):
+                        value_str_time = (
+                            evaluate_relative_date(condition.value)
+                            .astimezone()
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:726]: This converts the TZ to the system time, and forces the TZ on naive objects. I thought we agreed on UTC ?

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 11]
```diff
@@ -683,10 +718,33 @@ def _get_single_condition_criterion(
...
+                            .strftime("%Y-%m-%d %H:%M:%S")
+                        )
+                    elif isinstance(condition.value, datetime):
+                        value_str_time = condition.value.astimezone().strftime("%Y-%m-%d %H:%M:%S")
+                    else:
+                        from dateutil import parser as dateutil_parser
+
+                        value_str_time = (
+                            dateutil_parser.parse(condition.value)
+                            .astimezone()
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:736]: This converts the TZ to the system time, and forces the TZ on naive objects. I thought we agreed on UTC ?

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 12]
```diff
@@ -683,10 +718,33 @@ def _get_single_condition_criterion(
...
+
+                if isinstance(condition.value, (RelativeDate, datetime, str)):
+                    if isinstance(condition.value, RelativeDate):
+                        value_str_time = (
+                            evaluate_relative_date(condition.value)
+                            .astimezone()
+                            .strftime("%Y-%m-%d %H:%M:%S")
+                        )
+                    elif isinstance(condition.value, datetime):
+                        value_str_time = condition.value.astimezone().strftime("%Y-%m-%d %H:%M:%S")
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:730]: This converts the TZ to the system time, and forces the TZ on naive objects. I thought we agreed on UTC ?

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 10 above)
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py:726]: I may not getting well these TZ manipulation but, adding as argument `timezone.utc` to astimezone, force me to change fixtures tests, otherwise tests will not work anymore and we should not change fixture tests !

That's why i remove that parameter !

These are tests failing !
```
tests/backends/sql_translator_integration_tests/test_sql_postgres_pypika_translator_steps.py::test_sql_translator_pipeline[/filter_dates_from_until_pypika]
FAILED tests/backends/sql_translator_integration_tests/test_sql_postgres_pypika_translator_steps.py::test_sql_translator_pipeline[/filter_relative_dates_pypika]
FAILED tests/backends/sql_translator_integration_tests/test_sql_postgres_pypika_translator_steps.py::test_sql_translator_pipeline[/ifthenelse_dates_conditions_pypika]
```

Can you be more precise ?

**Review by lukapeschke — APPROVED** [review]
:medal_sports:
